Lib/asyncio/streams.py

```
@@ -802,3 +1042,675 @@ def __aiter__(self):
+        if self._buffer:
+            info.append(f'{len(self._buffer)} bytes')
+        if self._eof:
+            info.append('eof')
+        if self._limit != _DEFAULT_LIMIT:
+            info.append(f'limit={self._limit}')
+        if self._waiter:
+            info.append(f'waiter={self._waiter!r}')
+        if self._exception:
+            info.append(f'exception={self._exception!r}')
+        if self._transport:
+            info.append(f'transport={self._transport!r}')
+        if self._paused:
+            info.append('paused')
+        return '<{}>'.format(' '.join(info))
+
+    @property
+    def mode(self):
+        return self._mode
+
+    def is_server_side(self):
+        return self._is_server_side
+
+    @property
+    def transport(self):
+        return self._transport
+
+    def write(self, data):
+        self._mode._check_write()
+        self._transport.write(data)
+        return self._fast_drain()
+
+    def writelines(self, data):
+        self._mode._check_write()
+        self._transport.writelines(data)
+        return self._fast_drain()
+
+    def _fast_drain(self):
+        # The helper tries to use fast-path to return already existing
+        # complete future object if underlying transport is not paused
+        #and actual waiting for writing resume is not needed
+        exc = self.exception()
+        if exc is not None:
+            fut = self._loop.create_future()
+            fut.set_exception(exc)
+            return fut
+        if not self._transport.is_closing():
+            if self._protocol._connection_lost:
+                fut = self._loop.create_future()
+                fut.set_exception(ConnectionResetError('Connection lost'))
+                return fut
+            if not self._protocol._paused:
+                # fast path, the stream is not paused
+                # no need to wait for resume signal
+                return self._complete_fut
+        return _OptionalAwait(self.drain)
```

I replaced returning `loop.create_task(drain())` with `_OptionalAwait(self.drain)`.
The reason is: I don't like that `drain()` code is executed even if user didn't ask (plain `stream.write(data)` call).
If `drain()` failed with exception but `write()` was not awaited the exception was discarded.

It works pretty well I believe but confuses code readers. Also, future modifications of `drain()` method can produce a weird side-effect.
Better to avoid execution of dead code at all, the mental model becomes much clearer at least.